examples/mnit_ballz/src/game_logic.nit

```diff
@@ -14,167 +14,190 @@
+				intersecting_lines.add(l)
+			end
 		end
 
-		# y value
-		if ny >= 0.0 and ny <= gh then
-			self.y = ny
-		else if ny < 0.0 then
-			if not walls_activated then self.y = gh else do_bounce(2)
-		else if ny > gh then
-			if not walls_activated then self.y = 0.0 else do_bounce(2)
+		# get the line closest to the ball from the intersecting lines, setting the closest point
+		var min_dist = 100.0
+		if intersecting_lines.length >= 2 then
+			for l in intersecting_lines do
+				var closest = point_on_line_closest_to_ball(new_center, l.point_left, l.point_right)
+				var distance = distance(closest, new_center)
+				if distance < min_dist then
+					min_dist = distance
+					intersecting_line = l
+					closest_point = closest
+				end
+			end
+		else if intersecting_lines.length == 1 then
+			intersecting_line = intersecting_lines[0]
+			closest_point = point_on_line_closest_to_ball(new_center, intersecting_line.point_left, intersecting_line.point_right)
 		end
-	end
 
-	# bounce in function of the position of the wall relative to the ball: 1=left or right, 2=top or down
-	fun do_bounce(wall_position: Int)
-	do
-		if wall_position == 1 then
-			offset_x = -offset_x*0.85
-		else if wall_position == 2 then
-			offset_y = -offset_y*0.85
-		end
-		if offset_x.abs > 1.0 and offset_y.abs > 1.0 then
-		self.x += offset_x
-		self.y += offset_y
+		if intersecting_line != null and closest_point != null then
+			return bounce(center, intersecting_line.point_left, intersecting_line.point_right, offset)
 		end
+		return null
 	end
 
-	fun activate_walls(event: MotionEvent)
-	do
-		if event.just_went_down then
-			walls_activated = not walls_activated
-		end
-	end
+	# Event interception
+	fun intercepts(event: InputEvent): Bool is abstract
 end
 
-class Screen
-	var ball_img: Image
-	var game: Game
+# The core of the game
+class Game
 
-	init(app: App, display: Display)
-	do
-		game = new Game(display)
-		ball_img = app.load_asset("images/ball.png").as(Image)
-		var scale = game.img_dim.to_f / game.img_ori_dim.to_f
-		ball_img.scale = scale
-		ball_img.scale = 3.0
-	end
+	# The Ball !
+	var ball: Ball is noinit
 
-	fun do_frame(display: Display)
-	do
-		display.clear(0.0, 0.0, 0.0)
-		display.blit_rotated(ball_img, game.ball.x, game.ball.y, 0.0)
-	end
+	# List of walls in the level
+	var walls: Array[Wall] is noinit
 
-	fun input(ie: InputEvent): Bool
-	do
-		if ie isa ASensorProximity then
-			if ie.distance == 0.0 then ball_img.scale = 6.0 else ball_img.scale = 3.0
-		else
-			game.ball.intercepts(ie)
-		end
-		return true
-	end
-end
+	# Width of the display
+	var width: Float is noinit
 
-class Game
-	var ball: Ball
-	var width: Float
-	var height: Float
+	# Heightof the display
+	var height: Float is noinit
 
-	var img_ori_dim: Int = 256
-	fun img_dim: Int do return 210
+	# Quadtree used for collision detection
+	var quadtree: SQuadTree[OrientedLine[Float]]
 
 	init(display: Display)
```

Comment:
Old init style is old.
And bad OO design as you add unneeded coupling Game -> Display here.
Use a single autoinit constructor by removing `noautoinit` from width and height.